cpp/src/text/jaccard.cu

```diff
@@ -0,0 +1,305 @@
+ * @return A device-span of the row values
+ */
+__device__ auto get_row(cudf::column_device_view const& d_input, cudf::size_type idx)
+{
+  auto const offsets =
+    d_input.child(cudf::lists_column_view::offsets_column_index).data<cudf::size_type>();
+  auto const offset = offsets[idx];
+  auto const size   = offsets[idx + 1] - offset;
+  auto const begin =
+    d_input.child(cudf::lists_column_view::child_column_index).data<uint32_t>() + offset;
+  return cudf::device_span<uint32_t const>(begin, size);
+}
+
+/**
+ * @brief Count the unique values within each row of the input column
+ *
+ * This is called with a warp per row
+ */
+struct jaccard_unique_fn {
+  cudf::column_device_view const d_input;
+  cudf::size_type* d_results;
+
+  // warp per row
+  __device__ void operator()(cudf::size_type idx) const
+  {
+    using warp_reduce = cub::WarpReduce<cudf::size_type>;
+    __shared__ typename warp_reduce::TempStorage temp_storage;
+
+    auto const row_idx  = idx / cudf::detail::warp_size;
+    auto const lane_idx = idx % cudf::detail::warp_size;
+    auto const row      = get_row(d_input, row_idx);
+    auto const begin    = row.begin();
+
+    cudf::size_type count = 0;
+    for (auto itr = begin + lane_idx; itr < row.end(); itr += cudf::detail::warp_size) {
+      count += (itr == begin || *itr != *(itr - 1));
+    }
+    auto const result = warp_reduce(temp_storage).Sum(count);
+    if (lane_idx == 0) { d_results[row_idx] = result; }
+  }
+};
+
+rmm::device_uvector<cudf::size_type> compute_unique_counts(cudf::column_view const& input,
+                                                           rmm::cuda_stream_view stream)
+{
+  auto const d_input = cudf::column_device_view::create(input, stream);
+  auto d_results     = rmm::device_uvector<cudf::size_type>(input.size(), stream);
+  jaccard_unique_fn fn{*d_input, d_results.data()};
+  thrust::for_each_n(rmm::exec_policy(stream),
+                     thrust::counting_iterator<cudf::size_type>(0),
+                     input.size() * cudf::detail::warp_size,
+                     fn);
+  return d_results;
+}
+
+/**
+ * @brief Count the number of common values within each row of the 2 input columns
+ *
+ * This is called with a warp per row
+ */
+struct jaccard_intersect_fn {
+  cudf::column_device_view const d_input1;
+  cudf::column_device_view const d_input2;
+  cudf::size_type* d_results;
+
+  // warp per row
+  __device__ float operator()(cudf::size_type idx) const
+  {
+    using warp_reduce = cub::WarpReduce<cudf::size_type>;
+    __shared__ typename warp_reduce::TempStorage temp_storage;
+
+    auto const row_idx  = idx / cudf::detail::warp_size;
+    auto const lane_idx = idx % cudf::detail::warp_size;
+
+    auto const needles  = get_row(d_input1, row_idx);
+    auto const haystack = get_row(d_input2, row_idx);
+
+    auto begin     = haystack.begin();
+    auto const end = haystack.end();
+
+    // TODO: investigate cuCollections device-side static-map to match row values
```

Comment:
When investigating this, it would also be worth considering whether it would be optimal to select which column is the needles and which is the haystack based on the size of the row (I would guess you'd want to build the hash table on the larger row and search with the smaller row, but it may not really matter that much).